src/pandas/types/numeric.h

```diff
@@ -44,14 +106,153 @@ class PANDAS_EXPORT IntegerArray : public NumericArray<TYPE> {
+    // would first create the valid_bits_ if needed
+    BitUtil::ClearBit(
+        static_cast<MutableBuffer&>(*valid_bits_).mutable_data(), this->offset_ + i);
+  }
+
+  // Returns whether or not the value at the specified index has
+  // been marked as null.
+  bool IsNull(int64_t i) const {
+    if (valid_bits_) {
+      // TODO: range check?
+      return BitUtil::GetBit(valid_bits_->data(), this->offset_ + i);
+    } else {
+      return false;
+    }
+  }
+
+  // Copy the list of nulls (aka valid bits) to a buffer
+  Status CopyNulls(int64_t length, std::shared_ptr<Buffer>* out) const {
+    // Pre-condition is that we already have valid_bits_ set
+    return valid_bits_->Copy(this->offset_, BitUtil::BytesForBits(length), out);
+  }
+
+  template <typename OTHER_TYPE>
+  IntegerArray<TYPE>& operator+=(const IntegerArray<OTHER_TYPE>& other) {
+    // TODO: check Status and throw exception
+    EnsureMutable();
+    auto length = std::min(this->length_, other.length());
+    if (HasNulls()) {
+      // Perform a |= on the valid bits if the other array has them.
+      // TODO: do this more efficiently, perhaps by passing our
+      //       valid_bits_ like this: other.BitwiseOr(*valid_bits_).
+      //       It might also be useful to have a class that wraps
+      //       much of the valid bit logic.
+      if (other.HasNulls()) {
+        for (auto ii = 0; ii < length; ++ii) {
+          if (IsNull(ii) || other.IsNull(ii)) {
+            SetNull(ii);
+          } else {
+            SetValid(ii);
+          }
+        }
+      }
+    } else if (other.HasNulls()) {
+      std::shared_ptr<Buffer> new_valid_bits;
+      // TODO: check Status and throw an exception
+      other.CopyNulls(length, &new_valid_bits);
+    }
+    auto this_data = this->mutable_data();
+    auto other_data = other.data();
+    if (HasNulls()) {
+      for (auto ii = 0; ii < length; ++ii, ++this_data, ++other_data) {
+        if (!IsNull(ii)) { *this_data += *other_data; }
+      }
+    } else {
+      // Can just copy directly
+      // TODO: optimize with SIMD when TYPE::size == OTHER_TYPE::size
+      for (auto ii = 0; ii < length; ++ii, ++this_data, ++other_data) {
+        *this_data += *other_data;
+      }
+    }
+    return *this;
+  }
+
+  // Compile time logic for determining the result type of
+  // a division of two integers.
+  template <typename OTHER_TYPE>
+  class IntegerDivideInteger {
+   public:
+    static constexpr auto max_value =
+        (std::numeric_limits<typename TYPE::c_type>::max() >
+                    std::numeric_limits<typename OTHER_TYPE::c_type>::max()
+                ? std::numeric_limits<typename TYPE::c_type>::max()
+                : std::numeric_limits<typename OTHER_TYPE::c_type>::max());
+
+    static constexpr auto needs_double =
+        (max_value > std::numeric_limits<typename FloatType::c_type>::max());
+
+    using type = typename std::conditional<needs_double, DoubleType, FloatType>::type;
```

Comment:
We could support integer division with `//`, but that's certainly a lower priority.